### ISSUE TITLE
Implement network address sets

### ISSUE THREAD

**Issue by stgraber** [lxc/incus#1450]
One current missing piece of our networking puzzle is support for address sets.
That is, a named set of address and/or subnets which can be used as source/destination selector within ACLs.

There is a native OVN object for this feature, so it would make sense to initially implement this for OVN networking.

Address sets should be scoped the same way as ACLs as they'll be very closely linked. This means address sets should be part of a project when `features.networks=true`

The expected changes should roughly be:
 - Allocate API extension
 - New database table (schema change)
 - New database functions (using the generator)
 - Add REST API structs for network address sets
 - Add OVN NB bindings for managing address sets
 - Add REST API to manipulate address sets
 - Add integration with OVN ACLs
 - Implement CLI command
 - Refresh translations
 - Add tests

**Comment by irhndt** [comment]
Hello, I’m giving this a try, but I’m new to both Go and Incus development. I’m approaching this as an enthusiast and treating it as a learning exercise so you may want someone that is already proficient. I already have a few questions.

In internal/server/network/ovn/ovn_nb_actions.go, I noticed some bindings for address sets. Am I correct in assuming that the ones we need to implement are in driver_ovn.go?

Additionally, I’m currently exploring where to implement the REST API and believe I’ve identified the relevant structures for API definitions. To summarize, I’m focusing on implementing:

    OVN NB bindings for managing address sets
    A REST API to manipulate address sets


Any guidance would be appreciated.

**Comment by stgraber** [comment]
Oh, I completely forgot that we were already making use of address sets internally for the `@internal` and `@external` selectors in our ACLs, so that indeed means that the work in `ovn_nb_actions.go` shouldn't be necessary as those functions already exist.

**Comment by stgraber** [comment]
@irhndt I assigned this issue to you so we don't end up with someone else working on this in parallel. Let me know if you have any questions as you start digging into this one!

**Comment by irhndt** [comment]
@stgraber  Thank you for assigning me this, I'll do my best.

I understand a bit better the code structure but still it's a huge and complex project <3.
I came with the following roadmap to help me focus and organize. However I still have some doubts on the right way to implement things and mostly where. I already worked on a few other files not mentioned in the roadmap such as `internal/version/api.go` or `internal/lifecycle/network_address_set.go`, mainly by grepping acl to find related files and assessing if I need a similar pattern for address sets.

I hope to begin tackle the main logic in the next few days, so before I take the big dive I would like to know if there is some precision for the phase 3 of my roadmap I should know before starting. Also if you spot some obvious mistakes or missing parts from previous phases let me know.

1. **Database Schema:**
    
    - [x]  Defined `address_sets` table in the fresh schema.
    - [x]  Added `updateFromV75` migration function.
    - [x]  Updated the `updates` map with version **76**.
    - [x]  Incremented `SchemaVersion` to **76**.
    - [x]  Ran `make update-schema` to regenerate `schema.go`.
2. **Database Layer:**
    
    - [x]  Created the `AddressSet` struct.
    - [x]  Implemented CRUD operations in `address_sets.go`.
    
3. **API and Logic Layer:**
    
    - [x]  Defined API structures in `shared/api/network_address_set.go`.
    - [ ]  Implemented handlers in  `internal/server/network/address_sets/`.
    - [ ]  Implement driver integration in `internal/server/network/driver_*`
    - [ ] Implement firewall integration either by a smart acl integration or direclty in `internal/server/firewall` looking forward to understand how ACL are used outside of OVN, also looking to understand how firewall backend integrates with OVN own filtering.
    
4. **OVN Integration:**
    - [ ]  Ensured OVN NBCTL supports required operations.
    
5. **CLI Integration:**
    
    - [ ]  Implemented CLI commands in `cmd/incus/address-set.go`.
    - [ ]  Provided help texts and usage examples.
    
6. **Translations:**
    
    - [ ]  Updated translation files with address set-related messages.
    - [ ]  Ensured translations are available in all supported languages.
    
7. **Testing:**
    
    - [ ]  Wrote unit tests for database operations.
    - [ ]  Wrote tests for API interactions.
    - [ ]  Wrote tests for CLI commands.
    - [ ]  Verified OVN integration through testing.
    
8. **Documentation:**
    Think about doc/network-addr-sets.md
    - [ ]  Updated API documentation in `doc/reference` and `doc/manpages`.
    - [ ]  Created user guides in `doc/howto`.

9. **Review and Validation:**
    
    - [ ]  Conducted code reviews for all new additions.
    - [ ]  Performed security audits on the new feature.

**Comment by stgraber** [comment]
In 2), note that we also generate as much of the database access functions as possible.
This isn't implemented for every object so depending on what you used as an example you may have missed it.

`network_integrations.go` would be a good place to look for another network related feature which does have code generated database functions.

**Comment by irhndt** [comment]
Hi hopes that everyone who reads this is well. I might need some help to run tests, I did not manage to find documentation about how to launch them or write them. It would help me a lot if somebody has some hints. I started by manually running main.sh in test/ however I guess it is not the way to go as it leaves me with a bunch of errors.

**Comment by irhndt** [comment]
Just a heads up, I got back on the development of the feature lately. For now I am missing:

- Enhanced testing for nftables (basics works)
- Testing for ovn
- Documentation

There should be some errors, but the feature is comming.

For the notice I made the choice to be able to use IP elements such as address and ranges in conjunction of address set in source / destination of ACLs. I will see if I can do the same for OVN while I test it.

**Comment by stgraber** [comment]
> I also have seen that a PR is ongoing in order to add filters for API collection, I may add it to address sets if necessary an advice ?

Yeah, we have the filter support everywhere now, so we'd want it for the address sets too. Should be a pretty trivial addition though.

**Comment by stgraber** [comment]
> Also I made a quite thorough script for testing address sets but I still haven't been able to integrate those in incus test suite (because I still don't have a success launching them 😄 ) I may need advice later on how to integrate them to test suite.

If you can't get the tests running locally, you can always use a draft PR so they run on Github instead. Or you can enable Github Actions in your fork on Github and have them run there whenever you push (even before opening a PR).

**Comment by stgraber** [comment]
> However the icmpv6 packet is not blocked.

I know we have some hardcoded default ACL entries that may be at play here and will allow at least a bunch of ICMP6 types, maybe that's interfering somehow?

**Comment by stgraber** [comment]
> This seems unwanted but as I am a beginner in incus and go development I am not sure if it is because of my code. If somebody knows something about it, a feedback would be appreciated.

ACLs default to reject, so the moment you apply an ACL to an instance or network, all traffic ingress and egress defaults to rejected. Traffic within the network itself also can get blocked.
Then anything you allow in the ACL gets allowed.

So I think that's what you're seeing. The two rejects are for default ingress/egress, the drop is for instance to instance within the network and then the accept is for what you added.

**Comment by irhndt** [comment]
Okay thank you for the feedback, so I will assess whether or not pre-existing rules exist

**Comment by irhndt** [comment]
> > This seems unwanted but as I am a beginner in incus and go development I am not sure if it is because of my code. If somebody knows something about it, a feedback would be appreciated.
> 
> ACLs default to reject, so the moment you apply an ACL to an instance or network, all traffic ingress and egress defaults to rejected. Traffic within the network itself also can get blocked. Then anything you allow in the ACL gets allowed.
> 
> So I think that's what you're seeing. The two rejects are for default ingress/egress, the drop is for instance to instance within the network and then the accept is for what you added.

I have adapted tests to use allow ACL and the behaviour is ok

**Comment by irhndt** [comment]
> > However the icmpv6 packet is not blocked.
> 
> I know we have some hardcoded default ACL entries that may be at play here and will allow at least a bunch of ICMP6 types, maybe that's interfering somehow?

I am wondering this is not about the feature I am currently working on and maybe we want to address this in another PR? This would imply some changes in base nftables rules templates I think. From what I understood it lies in jump rules, somehow for icmpv4 we jump to acl.incusbr0 and for icmpv6 it never happens. 
The full trace :

```go
IPV6

root@nexus-dev:/home/developer# python3 nftables-tracer.py "ip6 nexthdr ipv6-icmp"
trace id 8b77bb2d inet incus out.incusbr0 packet: oif "incusbr0" ip6 saddr fd42:ee1c:3a66:1e67::1 ip6 daddr fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc ip6 dscp cs0 ip6 ecn not-ect ip6 hoplimit 64 ip6 flowlabel 635295 ip6 nexthdr ipv6-icmp ip6 length 64 icmpv6 type echo-request icmpv6 code 0 icmpv6 parameter-problem 73138177 icmpv6 taddr ::4554:c367:0:0:6d16:300
trace id 8b77bb2d inet incus out.incusbr0 rule oifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, echo-request, nd-router-advert, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept (verdict accept)
trace id 8b77bb2d inet incus aclout.incusbr0 packet: oif "incusbr0" ip6 saddr fd42:ee1c:3a66:1e67::1 ip6 daddr fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc ip6 dscp cs0 ip6 ecn not-ect ip6 hoplimit 64 ip6 flowlabel 635295 ip6 nexthdr ipv6-icmp ip6 length 64 icmpv6 type echo-request icmpv6 code 0 icmpv6 parameter-problem 73138177 icmpv6 taddr ::4554:c367:0:0:6d16:300
trace id 8b77bb2d inet incus aclout.incusbr0 rule oifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, echo-request, nd-router-advert, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept (verdict accept)
trace id 8b77bb2d inet incus pstrt.incusbr0 packet: oif "incusbr0" ip6 saddr fd42:ee1c:3a66:1e67::1 ip6 daddr fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc ip6 dscp cs0 ip6 ecn not-ect ip6 hoplimit 64 ip6 flowlabel 635295 ip6 nexthdr ipv6-icmp ip6 length 64 icmpv6 type echo-request icmpv6 code 0 icmpv6 parameter-problem 73138177 icmpv6 taddr ::4554:c367:0:0:6d16:300
trace id 8b77bb2d inet incus pstrt.incusbr0 policy accept
trace id be85a15e inet incus in.incusbr0 packet: iif "incusbr0" ether saddr 00:16:3e:a2:6d:fc ether daddr 00:16:3e:3f:e3:53 ip6 saddr fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc ip6 daddr fd42:ee1c:3a66:1e67::1 ip6 dscp cs0 ip6 ecn not-ect ip6 hoplimit 64 ip6 flowlabel 246990 ip6 nexthdr ipv6-icmp ip6 length 64 icmpv6 type echo-reply icmpv6 code 0 icmpv6 parameter-problem 73138177 icmpv6 taddr ::4554:c367:0:0:6d16:300
trace id be85a15e inet incus in.incusbr0 policy accept
trace id be85a15e inet incus aclin.incusbr0 packet: iif "incusbr0" ether saddr 00:16:3e:a2:6d:fc ether daddr 00:16:3e:3f:e3:53 ip6 saddr fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc ip6 daddr fd42:ee1c:3a66:1e67::1 ip6 dscp cs0 ip6 ecn not-ect ip6 hoplimit 64 ip6 flowlabel 246990 ip6 nexthdr ipv6-icmp ip6 length 64 icmpv6 type echo-reply icmpv6 code 0 icmpv6 parameter-problem 73138177 icmpv6 taddr ::4554:c367:0:0:6d16:300
trace id be85a15e inet incus aclin.incusbr0 rule iifname "incusbr0" jump acl.incusbr0 (verdict jump acl.incusbr0)
trace id be85a15e inet incus acl.incusbr0 rule ct state established,related accept (verdict accept)



IPV4

root@nexus-dev:/home/developer# python3 nftables-tracer.py "ip protocol icmp"
trace id b1572690 inet incus out.incusbr0 packet: oif "incusbr0" ip saddr 10.16.93.1 ip daddr 10.16.93.110 ip dscp cs0 ip ecn not-ect ip ttl 64 ip id 46047 ip protocol icmp ip length 84 icmp type echo-request icmp code 0 icmp id 1117 icmp sequence 1
trace id b1572690 inet incus out.incusbr0 policy accept
trace id b1572690 inet incus aclout.incusbr0 packet: oif "incusbr0" ip saddr 10.16.93.1 ip daddr 10.16.93.110 ip dscp cs0 ip ecn not-ect ip ttl 64 ip id 46047 ip protocol icmp ip length 84 icmp type echo-request icmp code 0 icmp id 1117 icmp sequence 1
trace id b1572690 inet incus aclout.incusbr0 rule oifname "incusbr0" jump acl.incusbr0 (verdict jump acl.incusbr0)
trace id 24702fca inet incus out.incusbr0 packet: oif "lo" ip saddr 10.16.93.1 ip daddr 10.16.93.1 ip dscp cs6 ip ecn not-ect ip ttl 64 ip id 58245 ip protocol icmp ip length 112 icmp type destination-unreachable icmp code 3 icmp id 0 icmp sequence 0
trace id 24702fca inet incus out.incusbr0 policy accept
trace id 24702fca inet incus aclout.incusbr0 packet: oif "lo" ip saddr 10.16.93.1 ip daddr 10.16.93.1 ip dscp cs6 ip ecn not-ect ip ttl 64 ip id 58245 ip protocol icmp ip length 112 icmp type destination-unreachable icmp code 3 icmp id 0 icmp sequence 0
trace id 24702fca inet incus aclout.incusbr0 policy accept
trace id 53683fb6 inet incus in.incusbr0 packet: iif "lo" ip saddr 10.16.93.1 ip daddr 10.16.93.1 ip dscp cs6 ip ecn not-ect ip ttl 64 ip id 58245 ip protocol icmp ip length 112 icmp type destination-unreachable icmp code 3 icmp id 0 icmp sequence 0
trace id 53683fb6 inet incus in.incusbr0 policy accept
trace id 53683fb6 inet incus aclin.incusbr0 packet: iif "lo" ip saddr 10.16.93.1 ip daddr 10.16.93.1 ip dscp cs6 ip ecn not-ect ip ttl 64 ip id 58245 ip protocol icmp ip length 112 icmp type destination-unreachable icmp code 3 icmp id 0 icmp sequence 0
trace id 53683fb6 inet incus aclin.incusbr0 policy accept
trace id b1572690 inet incus acl.incusbr0 rule oifname "incusbr0" ip daddr @testAS_ipv4 ip protocol icmp reject with icmp port-unreachable (verdict drop)


nft list ruleset
table inet incus {
	set testAS_ipv4 {
		type ipv4_addr
		flags interval
		elements = { 10.16.93.110 }
	}

	set testAS_ipv6 {
		type ipv6_addr
		flags interval
		elements = { fd42:ee1c:3a66:1e67:216:3eff:fea2:6dfc }
	}

	chain acl.incusbr0 {
		ct state established,related accept
		oifname "incusbr0" ip daddr @testAS_ipv4 ip protocol icmp reject with icmp port-unreachable
		oifname "incusbr0" ip6 daddr @testAS_ipv6 ip6 nexthdr ipv6-icmp reject with icmpv6 port-unreachable
	}

	chain aclin.incusbr0 {
		type filter hook input priority filter; policy accept;
		iifname "incusbr0" tcp dport 53 accept
		iifname "incusbr0" udp dport 53 accept
		iifname "incusbr0" udp dport 67 accept
		iifname "incusbr0" udp dport 547 accept
		iifname "incusbr0" icmp type { destination-unreachable, time-exceeded, parameter-problem } accept
		iifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, nd-router-solicit, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept
		iifname "incusbr0" jump acl.incusbr0
	}

	chain aclout.incusbr0 {
		type filter hook output priority filter; policy accept;
		oifname "incusbr0" udp sport 67 accept
		oifname "incusbr0" udp sport 547 accept
		oifname "incusbr0" icmp type { destination-unreachable, time-exceeded, parameter-problem } accept
		oifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, echo-request, nd-router-advert, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept
		oifname "incusbr0" jump acl.incusbr0
	}

	chain aclfwd.incusbr0 {
		type filter hook forward priority filter; policy accept;
		iifname "incusbr0" jump acl.incusbr0
		oifname "incusbr0" jump acl.incusbr0
	}

	chain pstrt.incusbr0 {
		type nat hook postrouting priority srcnat; policy accept;
		ip saddr 10.16.93.0/24 ip daddr != 10.16.93.0/24 masquerade
		ip6 saddr fd42:ee1c:3a66:1e67::/64 ip6 daddr != fd42:ee1c:3a66:1e67::/64 masquerade
	}

	chain fwd.incusbr0 {
		type filter hook forward priority filter; policy accept;
		ip version 4 oifname "incusbr0" accept
		ip version 4 iifname "incusbr0" accept
		ip6 version 6 oifname "incusbr0" accept
		ip6 version 6 iifname "incusbr0" accept
	}

	chain in.incusbr0 {
		type filter hook input priority filter; policy accept;
		iifname "incusbr0" tcp dport 53 accept
		iifname "incusbr0" udp dport 53 accept
		iifname "incusbr0" icmp type { destination-unreachable, time-exceeded, parameter-problem } accept
		iifname "incusbr0" udp dport 67 accept
		iifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, nd-router-solicit, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept
		iifname "incusbr0" udp dport 547 accept
	}

	chain out.incusbr0 {
		type filter hook output priority filter; policy accept;
		oifname "incusbr0" tcp sport 53 accept
		oifname "incusbr0" udp sport 53 accept
		oifname "incusbr0" icmp type { destination-unreachable, time-exceeded, parameter-problem } accept
		oifname "incusbr0" udp sport 67 accept
		oifname "incusbr0" icmpv6 type { destination-unreachable, packet-too-big, time-exceeded, parameter-problem, echo-request, nd-router-advert, nd-neighbor-solicit, nd-neighbor-advert, mld2-listener-report } accept
		oifname "incusbr0" udp sport 547 accept
	}
}
```

**Comment by irhndt** [comment]
@stgraber 

Okay so I'll write documentation tomorrow, do a bit of cleaning and if you validate that we don't care for now to this icmpv6 blocking issue, I will try to merge in my main branch using github actions. Then I 'll do a PR.
By the way, in the script I use for my tests I rely on containers, acls and ovn actively to test behaviour on pings and tcp in every case that made sense for me, but I do not see such extended checks in for example `test/suites/network_acl.sh`, for now I have commited full tests in my branch but I do not know if we want to keep them.
